src/lib/y2network/connection_config/base.rb

```diff
@@ -62,6 +64,15 @@ def type
         const_name = self.class.name.split("::").last.upcase
         InterfaceType.const_get(const_name)
       end
+
+      # @param interfaces [Y2Network::InterfacesCollection]
+      def update_interfaces!(interfaces)
+        nil
+      end
+
+      def virtual?
```

Comment:
documentation

Reply:
As said before, virtual connection mainly means a 1:1 relation with virtual interfaces and we need a way to differentiate from fake interfaces.

Reply:
to be honest I'm still lost in current sense of Interface classes. 

Existing device (i mean real NIC e.g. ethernet one) -> physical_interface - I'm fine with that. We currently don't "need" it but i can see it's purpose

bond / bridge / vlan and other beasts -> virtual_interface - why? It is configuration stuff and in theory it can easily change when a config is switched to another profile

non existing device (a configuration for a physical device which is not present) - currently can be also seen as a configuration stuff as we don't handle hotplug events and do not re-read interfaces during the module's life so it never can turn to physical interface.

So as I see it now we need just one class for physical interfaces (which is currently not much useful at this time anyway) - and keeping it bcs of faked renaming example (deleting existing virtual config and creating new one) is not a reason for keeping bunch of code living.

If i summarize it. If you want to detect connection type you can check whether it has an interface assigned. If yes then it is a physical one. If not then it is either virtual or fake. If you are able to detect virtual type (bond, br, ...) from the config you have a virtual connection. If not you have a fake one.

Reply:
> to be honest I'm still lost in current sense of Interface classes.
> 
> Existing device (i mean real NIC e.g. ethernet one) -> physical_interface - I'm fine with that. We currently don't "need" it but i can see it's purpose
> 
> bond / bridge / vlan and other beasts -> virtual_interface - why? It is configuration stuff and in theory it can easily change when a config is switched to another profile

So, we can avoid adding them to the list of interfaces, just to not declare all used by existent configurations but only define the virtual interface name.

> 
> non existing device (a configuration for a physical device which is not present) - currently can be also seen as a configuration stuff as we don't handle hotplug events and do not re-read interfaces during the module's life so it never can turn to physical interface.

Agree
> 
> So as I see it now we need just one class for physical interfaces (which is currently not much useful at this time anyway) - and keeping it bcs of faked renaming example (deleting existing virtual config and creating new one) is not a reason for keeping bunch of code living.

Well, them are abstractions that are currently valid and the minimal we thought that could make sense to add. For example you could read information about udev rules and has always associated an interface. A physical one in case of present or a fake in case of not.

When read you could also represent virtual interfaces that are currently listed but not managed but that is other history. 

The main point for me is how and what to associate with a connection.

> 
> If i summarize it. If you want to detect connection type you can check whether it has an interface assigned. If yes then it is a physical one. If not then it is either virtual or fake. If you are able to detect virtual type (bond, br, ...) from the config you have a virtual connection. If not you have a fake one.

Reply:
I will try to summarize the rationale behind interfaces and connection configurations concepts. The idea of the interfaces class is to represent real system interfaces (for real, I mean interfaces under `/sys/class/net`, not only physical ones). The exception is the `FakeInterface` class that are interfaces which we *know* that will exist at some point (USB, hotplug, etc.).

Connection configurations are responsible to describe interfaces configurations. We usually consider a 1:1 relationship with interfaces, but it is not limited to that. For instance, in NM you can specify different profiles for a given interface (it may depend if you are connecting at home or at work, using different wifi networks, etc.). So, basically, these are different from the interfaces. In sysconfig, they usually correspond to `ifcfg-*` files.

Having said that, we can of course adapt or modify those concepts, but I would say that the idea will allow us to represent regular ifcfg-* files, NM profiles or, in the future, Wicked roles.

Reply:
I understood the concept (at least I thought so) ... what i wanted to point was that we don't use the model (Interface hierarchy). Only case I currently see in the near future is udev for phy devices.

But then I saw a PRs for Conection config for bond / br where fake interfaces are created according to br / bond configuration and I've got lost as it breaks the concept above - bcs fake interfaces are created for ifcfgs not assigned to physical device. But in this case it was created bcs it could be a non-existent BRIDGE_PORT (e.g.)

So not to enlarge it ... I currently don't see a point in having interfaces hierarchy and insist on mapping of each connection config to an interface. I'm slightly afraid that we got hijacked and over engineering a bit

Reply:
> I understood the concept (at least I thought so) ... what i wanted to point was that we don't use the model (Interface hierarchy). Only case I currently see in the near future is udev for phy devices.
> 
> But then I saw a PRs for Conection config for bond / br where fake interfaces are created according to br / bond configuration and I've got lost as it breaks the concept above - bcs fake interfaces are created for ifcfgs not assigned to physical device. But in this case it was created bcs it could be a non-existent BRIDGE_PORT (e.g.)

I see your point and reasons behind it. Basically we wanted to have in the `Y2Network::Config` all the interfaces that were somehow managed by YaST. That is, physical, virtual or fake interface, but of course, we could avoid doing that just based on connection relations.

> 
> So not to enlarge it ... I currently don't see a point in having interfaces hierarchy and insist on mapping of each connection config to an interface. I'm slightly afraid that we got hijacked and over engineering a bit